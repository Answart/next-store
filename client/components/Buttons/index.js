export { AddToCart } from './AddToCart';
export { CheckoutCart } from './CheckoutCart';
export { DeleteProduct } from './DeleteProduct';
export { DeleteProductVariant } from './DeleteProductVariant';
export { Logout } from './Logout';
export { RemoveFromCart } from './RemoveFromCart';
export { RequestPasswordReset } from './RequestPasswordReset';
export { ToggleCart } from './ToggleCart';
export { UpdateCartItem } from './UpdateCartItem';
export { UpdatePermissions } from './UpdatePermissions';
