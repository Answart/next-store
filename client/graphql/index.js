export {
  TOGGLE_LOCAL_CARTOPEN_MUTATION,
  SIGNUP_MUTATION,
  SIGNIN_MUTATION,
  SIGNOUT_MUTATION,
  REQUEST_PASSWORD_RESET_MUTATION,
  RESET_PASSWORD_MUTATION,
  UPDATE_PERMISSIONS_MUTATION,
  CREATE_IMAGE_MUTATION,
  CREATE_PRODUCT_MUTATION,
  UPDATE_PRODUCT_MUTATION,
  DELETE_PRODUCT_MUTATION,
  CREATE_PROD_VARIANT_MUTATION,
  UPDATE_PROD_VARIANT_MUTATION,
  DELETE_PROD_VARIANT_MUTATION,
  ADD_TO_CART_MUTATION,
  UPDATE_CARTITEM_MUTATION,
  REMOVE_FROM_CART_MUTATION,
  CREATE_ORDER_MUTATION,
} from './Mutation';
export {
  LOCAL_CARTOPEN_QUERY,
  LOCAL_USER_QUERY,
  ALL_USERS_QUERY,
  CURRENT_USER_QUERY,
  ORDER_QUERY,
  ORDERS_QUERY,
  ORDER_ITEMS_QUERY,
  PRODUCT_QUERY,
  SHOP_PRODUCTS_QUERY,
  PAGINATION_QUERY,
} from './Query';
