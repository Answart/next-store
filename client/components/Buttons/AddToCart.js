import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { ADD_TO_CART_MUTATION, CURRENT_USER_QUERY } from '../../graphql';


const AddToCart = ({ variant, disabled }) => (
  <Mutation mutation={ADD_TO_CART_MUTATION}
    variables={{ id: variant.id }}
    refetchQueries={[{ query: CURRENT_USER_QUERY }]}
  >
    {(addToCart, { loading }) => (
      <button className="big-btn"
        disabled={!variant.id || disabled}
        onClick={(e) => {
          e.preventDefault();
          addToCart()
            .catch(err => alert(err.message.replace('GraphQL error: ', '')));
        }}
      >
        Add{loading ? 'ing' : ''} To Cart
      </button>
    )}
  </Mutation>
);

AddToCart.propTypes = {
  variant: PropTypes.shape({
    id: PropTypes.string.isRequired
  }).isRequired,
  disabled: PropTypes.bool.isRequired
};


export { AddToCart };
