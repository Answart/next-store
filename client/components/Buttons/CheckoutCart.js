import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import NProgress from 'nprogress';
import { Mutation } from 'react-apollo';
import { CREATE_ORDER_MUTATION, CURRENT_USER_QUERY, LOCAL_CARTOPEN_QUERY } from '../../graphql';


const CheckoutCart = props => {
  const onToken = async (e, tokenRes, createOrder) => {
    if (!!e && !!e.preventDefault) e.preventDefault();
    NProgress.start();
    await createOrder({ variables: { token: tokenRes.id }}).then((res) => {
      if (!res || !res.data) return NProgress.done();
      Router.push({
        pathname: '/order',
        query: { id: res.data.createOrder.id },
      });
    });
  };
  const update = (cache, payload) => {
    try {
      const data = cache.readQuery({ query: LOCAL_CARTOPEN_QUERY });
      data.cartOpen = !data.cartOpen;
      cache.writeQuery({ query: LOCAL_CARTOPEN_QUERY, data });
    } catch(e) {}
  }
  return (
    <Mutation mutation={CREATE_ORDER_MUTATION}
      refetchQueries={[{ query: CURRENT_USER_QUERY }]}
      update={update}
      onError={(e) => {
        NProgress.done();
        alert(e.message.replace('GraphQL error: ', ''));
      }}
    >
      {createOrder => (
          <button className="big-btn"
            disabled={props.disabled || !props.email}
            onClick={(e) => onToken(e, { id: 'dummy-token' }, createOrder)}
          >
            {props.children}
          </button>
      )}
    </Mutation>
  );
};

CheckoutCart.propTypes = {
  totalQuantity: PropTypes.number.isRequired,
  totalAmount: PropTypes.number.isRequired,
  email: PropTypes.string.isRequired,
  image_url: PropTypes.string.isRequired,
  disabled: PropTypes.bool.isRequired,
  children: PropTypes.string.isRequired
};


export { CheckoutCart };
