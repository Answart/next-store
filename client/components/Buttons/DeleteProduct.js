import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import { DELETE_PRODUCT_MUTATION, SHOP_PRODUCTS_QUERY } from '../../graphql';


const DeleteProduct = props => {
  const { id, userName, className, children } = props;
  const update = (cache, payload) => {
    try {
      const variables = { name: userName };
      const data = cache.readQuery({ query: SHOP_PRODUCTS_QUERY, variables });
      data.products = data.products.filter(product => product.id !== payload.data.deleteProduct.id);
      cache.writeQuery({ query: SHOP_PRODUCTS_QUERY, variables, data });
    } catch(e) {}
  }
  return (
    <Mutation mutation={DELETE_PRODUCT_MUTATION}
      variables={{ id }}
      update={update}
    >
      {(deleteProduct, { error }) => (
        <button className={className ? `${className}` : 'dlt-btn'}
          disabled={!id}
          onClick={() => {
            if (confirm("Are you sure you want to delete this product?")) {
              deleteProduct()
                .catch(err => alert(err.message))
                .then((res) => {
                  Router.push({
                    pathname: "/shop",
                    query: { name: userName }
                  });
                });
            }
          }}
        >
          {children}
        </button>
      )}
    </Mutation>
  )
};

DeleteProduct.propTypes = {
  id: PropTypes.string.isRequired,
  userName: PropTypes.string.isRequired,
  className: PropTypes.string,
  children: PropTypes.string.isRequired
};


export { DeleteProduct };
