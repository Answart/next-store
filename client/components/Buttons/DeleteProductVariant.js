import React from 'react';
import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { DELETE_PROD_VARIANT_MUTATION, PRODUCT_QUERY } from '../../graphql';


const DeleteProductVariant = props => {
  const { productId, id, children } = props;
  const update = (cache, payload) => {
    try {
      const variables = { id: productId };
      const data = cache.readQuery({ query: PRODUCT_QUERY, variables });
      data.product.variants = data.product.variants.filter(variant => variant.id !== payload.data.deleteProductVariant.id);
      cache.writeQuery({ query: PRODUCT_QUERY, variables, data });
    } catch(e) {}
  };
  return (
    <Mutation mutation={DELETE_PROD_VARIANT_MUTATION}
      variables={{ id }}
      update={update}
    >
      {(deleteProductVariant, { error }) => (
        <button className="dlt-btn"
          disabled={!id}
          onClick={() => {
            if (confirm("Are you sure you want to delete this selection?")) {
              deleteProductVariant()
                .catch(err => alert(err.message))
                .then((res) => {
                  if (props.postDelete) props.postDelete();
                });
            }
          }}
        >
          {children}
        </button>
      )}
    </Mutation>
  );
};

DeleteProductVariant.propTypes = {
  productId: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  children: PropTypes.string.isRequired
};


export { DeleteProductVariant };
