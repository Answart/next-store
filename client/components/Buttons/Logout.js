import React from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';
import { Mutation } from 'react-apollo';
import { SIGNOUT_MUTATION, CURRENT_USER_QUERY } from '../../graphql';


const Logout = () => (
  <Mutation mutation={SIGNOUT_MUTATION}
    refetchQueries={[{ query: CURRENT_USER_QUERY }]}
  >
    {(signout) => (
      <button className="undrln-btn"
        onClick={() => {
          signout()
            .catch(err => alert(err.message))
            .then((res) => {
              Router.push({ pathname: '/' });
            })
        }}
      >
        Log Out
      </button>
    )}
  </Mutation>
);


export { Logout };
