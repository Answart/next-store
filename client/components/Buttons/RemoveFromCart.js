import PropTypes from 'prop-types';
import { Mutation } from 'react-apollo';
import { REMOVE_FROM_CART_MUTATION, CURRENT_USER_QUERY } from '../../graphql';


const RemoveFromCart = ({ id }) => {
  const update = (cache, payload) => {
    try {
      const data = cache.readQuery({ query: CURRENT_USER_QUERY });
      const cartItemId = payload.data.removeFromCart.id;
      data.me.cart = data.me.cart.filter(cartItem => cartItem.id !== cartItemId);
      cache.writeQuery({ query: CURRENT_USER_QUERY, data });
    } catch(e) {}
  };
  return (
    <Mutation mutation={REMOVE_FROM_CART_MUTATION}
      variables={{ id }}
      update={update}
      optimisticResponse={{
        __typename: 'Mutation',
        removeFromCart: {
          __typename: 'CartItem',
          id
        }
      }}
    >
      {(removeFromCart, { loading, error }) => (
        <button className="undrln-btn"
          disabled={!id}
          onClick={(e) => {
            e.preventDefault();
            removeFromCart()
              .catch(err => alert(err.message.replace('GraphQL error: ', '')));
          }}
          title="Delete Item"
        >
          Remov{loading ? 'ing' : 'e'}
        </button>
      )}
    </Mutation>
  );
};

RemoveFromCart.propTypes = {
  id: PropTypes.string.isRequired
};


export { RemoveFromCart };
