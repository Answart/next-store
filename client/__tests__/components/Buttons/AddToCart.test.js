import wait from 'waait';
import toJSON from 'enzyme-to-json';
import { mount } from 'enzyme';
import { MockedProvider } from 'react-apollo/test-utils';
import { ApolloConsumer } from 'react-apollo';
import {
  mockVariant,
  userQueryEmptyCartMock, userQueryCartItemMock,
  addToCartMutationMock,
} from '../../../lib/test-utils/mocks';
import { AddToCart } from '../../../components/Buttons';
import { CURRENT_USER_QUERY } from '../../../graphql';

const mocks = [
  { ...userQueryEmptyCartMock },
  { ...addToCartMutationMock },
  userQueryCartItemMock({ quantity: 1 }),
];


describe('<AddToCart />', () => {
  afterAll(() => wrapper.unmount());
  afterEach(() => jest.clearAllMocks());

  it('renders and matches the snap shot', async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <AddToCart variant={mockVariant} disabled={false} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(toJSON(wrapper.find('button'))).toMatchSnapshot();
    wrapper.unmount();
  });

  it('adds an item to cart when clicked', async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return (<AddToCart variant={mockVariant} disabled={false} />);
          }}
        </ApolloConsumer>
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const { data: { me: { cart } } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(cart).toHaveLength(0);
    wrapper.find('button').simulate('click');
    await wait(50);
    const { data: { me: { cart: cart2 } } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(cart2).toHaveLength(1);
    expect(cart2[0].id).toBe('c4rt1t3m1d');
    expect(cart2[0].quantity).toBe(1);
    expect(cart2[0].variant.id).toBe('v4r13nt1d');
    expect(cart2[0].variant.quantity).toBe(4);
    wrapper.unmount();
  });

  it("changes from 'Add To Cart' to 'Adding To Cart' when clicked", async () => {
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <AddToCart variant={mockVariant} disabled={false} />
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    expect(wrapper.text()).toContain('Add To Cart');
    wrapper.find('button').simulate('click');
    expect(wrapper.text()).toContain('Adding To Cart');
    wrapper.unmount();
  });

  it('does not add an item to cart when disabled', async () => {
    let apolloClient;
    const wrapper = mount(
      <MockedProvider mocks={mocks}>
        <ApolloConsumer>
          {client => {
            apolloClient = client;
            return (<AddToCart variant={mockVariant} disabled={true} />);
          }}
        </ApolloConsumer>
      </MockedProvider>
    );
    await wait();
    wrapper.update();
    const { data: { me: { cart } } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(cart).toHaveLength(0);
    wrapper.find('button').simulate('click');
    await wait();
    expect(wrapper.text()).toContain('Add To Cart');
    const { data: { me: { cart: cart2 } } } = await apolloClient.query({ query: CURRENT_USER_QUERY });
    expect(cart2).toHaveLength(0);
    wrapper.unmount();
  });
});
