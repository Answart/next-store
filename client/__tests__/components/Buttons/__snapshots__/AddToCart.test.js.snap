// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AddToCart /> renders and matches the snap shot 1`] = `
<button
  className="big-btn"
  disabled={false}
  onClick={[Function]}
>
  Add
   To Cart
</button>
`;
